Document Java Source Context (#7086)
getsentry/sentry-docs: src/platform-includes/source-context/java.mdx

Reviewer: i'd like to proceed with approving and merging these changes, but I'd like us to consider updating these pages to work like the JS source maps pages today

- meaning a little less text
- similar image to show before and after 
- set up sub pages instead of lists of each option for plugin config

Ideally this should make the page easier to consume and things like UUIDs for debug files should be hidden away

I will take a crack at it later if I get time

Reviewer 2: It'd be awesome if we can split up the source code context in a product's page (explaining the feature but platform agnostic) so that platforms can just link to when they offer support.

This is tracked here https://www.notion.so/sentry/x-Source-Context-adoption-83a34f599c9745498aa20849ac44fdc8?pvs=4#43bdef3f5b2142e59b180fafae6e3bf8

Then the source code context could add includes (with code snippets) on how to install per platform, this makes the content/feature more discoverable and easy to add support.

    org = "___ORG_SLUG___"
    projectName = "___PROJECT_SLUG___"
    authToken = "your-sentry-auth-token"
}
```

```kotlin
buildscript {
    repositories {
        mavenCentral()
    }
}

plugins {
    id("io.sentry.jvm.gradle") version "{{@inject packages.version('sentry.java.android.gradle-plugin', '3.9.0') }}"
}

sentry {
    // Enables more detailed log output, e.g. for sentry-cli.
    //
    // Default is false.
    debug.set(true)

    // Generates a source bundle and uploads your source code to Sentry.
    // This enables source context, allowing you to see your source
    // code as part of your stack traces in Sentry.
    //
    // Default is disabled.
    includeSourceContext.set(true)

    // Includes additional source directories into the source bundle.
    // These directories are resolved relative to the project directory.
    additionalSourceDirsForSourceContext.set(setOf("mysrc/java", "other-source-dir/main/kotlin"))

    org.set("___ORG_SLUG___")
    projectName.set("___PROJECT_SLUG___")
    authToken.set("your-sentry-auth-token")
}
```

</PlatformSection>

<PlatformSection notSupported={["android"]}>

## Maven Plugin

> We also offer a [Sentry Maven Plugin](https://github.com/getsentry/sentry-maven-plugin).
> 
> You can add the plugin to your `pom.xml` by adding the following lines:

Reviewer 3: ```suggestion
Add the [Sentry Maven Plugin](https://github.com/getsentry/sentry-maven-plugin) to your project by adding the following lines to your `pom.xml` file:
```

                <!-- download the latest sentry-cli and provide path to it here -->
                <!-- download it here: https://github.com/getsentry/sentry-cli/releases -->
                <!-- minimum required version is 2.17.3 -->
                <sentryCliExecutablePath>/path/to/sentry-cli</sentryCliExecutablePath>

                <org>___ORG_SLUG___</org>

                <project>___PROJECT_SLUG___</project>

                <!-- in case you're self hosting, provide the URL here -->
                <!--<url>http://localhost:8000/</url>-->

                <!-- provide your auth token via SENTRY_AUTH_TOKEN environment variable -->
                <!-- you can find it in Sentry UI: Settings > Account > API > Auth Tokens -->
                <authToken>${env.SENTRY_AUTH_TOKEN}</authToken>
            </configuration>
            <executions>
                <execution>
                    <phase>install</phase>
                    <goals>
                        <goal>uploadSourceBundle</goal>
                    </goals>
                </execution>
            </executions>
        </plugin>
    </plugins>
    ...
</build>
```

> For now you have to manually download `sentry-cli` for your required architecture and point
> the Maven Plugin to it using `sentryCliExecutablePath`.
> You can get the latest release from the following URL:

Reviewer 3: ```suggestion
You must manually download `sentry-cli` for your required architecture and point
the Maven Plugin to it using `sentryCliExecutablePath`.
You can get the latest release from the following URL:
```

```
https://github.com/getsentry/sentry-cli/releases/tag/{{@inject apps.version('sentry-cli', '2.17.3') }}
```

</PlatformSection>

## Manually Uploading Source Context

> If your build tool of choice isn't supported yet or there are other reasonsfor not
> using our Gradle or Maven plugins, you can manually create and upload source bundles.
> 
> The `sentry-cli` commands allow you to supply `--org` and `--project` as well as providing the auth token by setting the `SENTRY_AUTH_TOKEN` environment variable.
> Another option is to use `.sentryclirc` or a `.properties` file which you can link using the `SENTRY_PROPERTIES` environment variable.

Reviewer 3: ```suggestion
If you're using a build tool we don't support, or you prefer not to use Sentry's build tool plugins, you'll need to create and upload source bundle files manually using the Sentry CLI. 

The `sentry-cli` commands allow you to supply `--org` and `--project` as well as provide the auth token by setting the `SENTRY_AUTH_TOKEN` environment variable.

You can also use a `.sentryclirc` or a `.properties` file, which you can link using the `SENTRY_PROPERTIES` environment variable.
```

### Creating the Source Bundle

> First you have to create the source bundle containing your source files.

Reviewer 3: ```suggestion
First, create the source bundle containing your source files:
```

```
sentry-cli debug-files bundle-jvm --output path/to/store/bundle --debug-id A_VALID_UUID path/to/source-code
```

### Uploading the Source Bundle

Next you can upload that source bundle to Sentry.

```
sentry-cli debug-files upload --type jvm path/to/bundle
```

### Configuring the SDK

> You have to tell the SDK which source bundle it should use for providing Source Context via one of the following options.

Reviewer 3: ```suggestion
You'll need to tell the SDK which source bundle it should use for providing Source Context via one of the following options:
```

#### `sentry-debug-meta.properties`

> Add a `sentry-debug-meta.properties` to your application resources at build time which will be picked up automatically by the SDK.

Reviewer 3: ```suggestion
Add a `sentry-debug-meta.properties` variable to your application resources at build time, which will be picked up automatically by the SDK:
```

the Sentry server to look up source code in the source bundle with a matching ID.

If you're using a build tool plugin, these steps happen automatically.

![Java Source Context](java-source-context.png)

All of the following methods require `org`, `project` and an `authToken`.

You can create an auth token by visiting the
[User Auth Tokens](https://sentry.io/settings/account/api/auth-tokens/) settings page in Sentry.io.
The auth token requires either a `project:releases` or `project:write` Scope.

## Known Limitations

- Files with same name but different extensions will lead to undefined behavior
  - e.g. MainActivity.java and MainActivity.kt will both be renamed to MainActivity.jvm
- Package declaration and file tree must match for source lookup to work
  - e.g. a class io.sentry.sample.MainActivity.java has to be stored in io/sentry/sample

<PlatformSection supported={["android"]}>

- Kotlin files are allowed to contain multiple classes but code obfuscation tools like ProGuard or R8 may prevent this from working correctly
- We don't add generated sources for AGP < 7.4

</PlatformSection>

## Using the Gradle Build Tool Plugin

> Add the Sentry Gradle plugin to your project by adding the following lines and making sure the `assemble` task is executed:

Author: Updated
